### PR TITLE
Upgrade to Polkadot v0.9.39

### PR DESCRIPTION
**Pull request by jasl**
https://github.com/paritytech/substrate/compare/polkadot-v0.9.37...polkadot-v0.9.38
https://github.com/paritytech/substrate/compare/polkadot-v0.9.38...polkadot-v0.9.39

Breaking changes from upstream:
- https://github.com/paritytech/subxt/pull/804
- https://github.com/paritytech/substrate/pull/13284
- https://github.com/paritytech/substrate/pull/12993

Other notable changes:
- https://github.com/paritytech/substrate/pull/13204
- https://github.com/paritytech/substrate/pull/12976
- https://github.com/paritytech/substrate/pull/13197
- https://github.com/paritytech/substrate/pull/13301
- https://github.com/paritytech/substrate/pull/12520

Notes:
`warning: the following packages contain code that will be rejected by a future version of Rust: nom v5.1.2`
which dependent by https://github.com/http-rs/surf (used by phactory)
it seems 1 year no update, maybe we should replace it

```
---- runtime::tests::gas_limit_works stdout ----
thread 'runtime::tests::gas_limit_works' panicked at 'events not registered at the genesis block', /Users/jasl/.cargo/git/checkouts/substrate-7e08433d4c370a21/1837f42/frame/system/src/lib.rs:1440:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- runtime::__construct_runtime_integrity_test::runtime_integrity_tests stdout ----
thread 'runtime::__construct_runtime_integrity_test::runtime_integrity_tests' panicked at 'Given `CallStack` height 6, `MaxCodeLen` should be set less than 82526 (current value is 2097152), to avoid possible runtime oom issues.', /Users/jasl/.cargo/git/checkouts/substrate-7e08433d4c370a21/1837f42/frame/contracts/src/lib.rs:396:13
```
relates to https://github.com/paritytech/substrate/pull/12993/files#diff-70e9723e9db62816e35f6f885b6770a8449c75a6c2733e9fa7a245fe52c4656cR423-R437

In adiition:
- Substrate main branch just upgrade grandpa to 0.16.2 https://github.com/paritytech/substrate/pull/13622 Do we need upgrade PRuntime V0 as well?

### CODE REVIEW

**Comment by kvinwang** [comment]
> `warning: the following packages contain code that will be rejected by a future version of Rust: nom v5.1.2`
> which dependent by https://github.com/http-rs/surf (used by phactory)

Removed.


> ---- runtime::tests::gas_limit_works stdout ----
> thread 'runtime::tests::gas_limit_works' panicked at 'events not registered at the genesis block', /Users/jasl/.cargo/git/checkouts/substrate-7e08433d4c370a21/1837f42/frame/system/src/lib.rs:1440:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Fixed.



> ---- runtime::__construct_runtime_integrity_test::runtime_integrity_tests stdout ----
> thread 'runtime::__construct_runtime_integrity_test::runtime_integrity_tests' panicked at 'Given `CallStack` height 6, `MaxCodeLen` should be set less than 82526 (current value is 2097152), to avoid possible runtime oom issues.', /Users/jasl/.cargo/git/checkouts/substrate-7e08433d4c370a21/1837f42/frame/contracts/src/lib.rs:396:13

A bit tricky for this one. Max size of 82526 is obviously not acceptable for phat contracts. They hard-coded a limitation in the tests and we can not change it. Let's make a workaround temporarily in order to pass the testing and discuss with parity for a better solution. I've left a message there on the parity's Element group. 

> * Substrate main branch just upgrade grandpa to 0.16.2 [Bump grandpa to 0.16.2 paritytech/substrate#13622](https://github.com/paritytech/substrate/pull/13622) Do we need upgrade PRuntime V0 as well?

No feature changes from 0.16.1 to 0.16.2. So I think we can ignore it.

File: crates/pink/src/runtime/pallet_pink.rs [hunk 1]
```diff
@@ -92,6 +92,12 @@ pub mod pallet {
             );
             UncheckedFrom::unchecked_from(<T as frame_system::Config>::Hashing::hash(&buf))
         }
+
+        fn deposit_address(contract_addr: &T::AccountId) -> T::AccountId {
+            let entropy = (b"contract_depo_v1", contract_addr).using_encoded(T::Hashing::hash);
+            Decode::decode(&mut TrailingZeroInput::new(entropy.as_ref()))
```
[the remark this answers is not in the document]
Comment (jasl) [reply crates/pink/src/runtime/pallet_pink.rs:98]: Could you help? I need check doc to know how to build a Vec for `unchecked_from`

[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink/src/runtime/pallet_pink.rs:98]:
```suggestion
            UncheckedFrom::unchecked_from(entropy)
```

File: crates/phala-node-rpc-ext/src/storage_changes.rs [hunk 2]
```diff
@@ -116,9 +116,12 @@ where
                     .map_err(|e| Error::invalid_block(id, e))?;
                 return Ok(StorageChanges {
                     main_storage_changes: state
-                        .pairs()
-                        .into_iter()
-                        .map(|(k, v)| (StorageKey(k), Some(StorageKey(v))))
+                        .pairs(Default::default())
+                        .unwrap()
+                        .map(|pair| {
+                            let (k, v) = pair.expect("Should get the key and value");
```
[the remark this answers is not in the document]
Comment (jasl) [reply crates/phala-node-rpc-ext/src/storage_changes.rs:122]:
```
fn pairs<'a>(&'a self, args: IterArgs) -> Result<PairsIter<'a, H, Self::RawIter>, Self::Error> {
		Ok(PairsIter {
			backend: Some(self),
			raw_iter: self.raw_iter(args)?,
			_phantom: Default::default(),
		})
	}
```

Get `PairIter` shall never fail, so `unwrap()` is safe, extracting KV from pair depends on backend logic, I just rough check ralates code

primitives/state-machine/src/trie_backend_essence.rs
```
let trie_iter = self
			.with_trie_db(root, args.child_info.as_ref(), |db| {
				let prefix = args.prefix.as_deref().unwrap_or(&[]);
				if let Some(start_at) = args.start_at {
					TrieDBRawIterator::new_prefixed_then_seek(db, prefix, &start_at)
				} else {
					TrieDBRawIterator::new_prefixed(db, prefix)
				}
			})
			.map_err(|e| format!("TrieDB iteration error: {}", e))?;
```

may fail

[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/phala-node-rpc-ext/src/storage_changes.rs:122]: > .map_err(|e| format!("TrieDB iteration error: {}", e))?;

It is acceptable to panic if some fault occurred in the underlying DB.

> Get PairIter shall never fail, so unwrap() is safe

Use expect should be better.
Except in testing or demo codes, I think we only use `.unwrap()` in those well-known contexts such as `some_mutex.lock().unwrap()`.

[the remark this answers is not in the document]
Comment (jasl) [reply crates/phala-node-rpc-ext/src/storage_changes.rs:122]: Ok I can update all unwrap() usage

[the remark this answers is not in the document]
Comment (jasl) [reply crates/phala-node-rpc-ext/src/storage_changes.rs:122]: fixed

**Review by kvinwang — APPROVED** [review]
LGTM

**Pull request merged**